### ISSUE TITLE
Quicksort is too slow (or merge sort is too fast)

### ISSUE THREAD

**Issue by mrjbq7** [slavapestov/factor#1357]
I'd love to have an in-place sort as part of the standard library, but not sure if quicksort is the right algorithm.  If we allow a "in-place" merge-sort which is really not in-place except that it copies the result back:

```factor
: merge-sort! ( seq -- )
    dup natural-sort swap 0 swap copy ; 

: foo ( seq -- n )
   [ 0 1000000 ] dip '[ _ dup merge-sort! length + ] times  ;

IN: scratchpad 100 iota >array gc [ foo ] time
Running time: 5.172995066 seconds
```

vs quicksort

```factor
: bar ( seq -- n )
   [ 0 1000000 ] dip '[ _ dup natural-sort! length + ] times  ;

IN: scratchpad 100 iota >array gc [ bar ] time
Running time: 11.208535713 seconds
```

It is a little distorted by resorting the same array (so the input is always sorted), but there is always a similar performance difference between the two.

**Comment by mrjbq7** [comment]
This makes it faster than merge-sort, although isn't something we should probably push it does array and fixnum specialization...

```diff
diff --git a/extra/sorting/quick/quick.factor b/extra/sorting/quick/quick.factor
index 38ee1b9..233c339 100644
--- a/extra/sorting/quick/quick.factor
+++ b/extra/sorting/quick/quick.factor
@@ -8,7 +8,10 @@ IN: sorting.quick
 
 <PRIVATE
 
+USING: arrays kernel.private math.private ;
+
 :: quicksort ( seq from to quot -- )
+    seq from to { array array-capacity array-capacity } declare 3drop
     from to < [
         from to + 2/ seq nth-unsafe :> pivot
 
@@ -16,16 +19,16 @@ IN: sorting.quick
             [
                 over seq nth-unsafe pivot quot call( x x -- x )
                 +lt+ eq?
-            ] [ [ 1 + ] dip ] while
+            ] [ [ 1 fixnum+fast ] dip ] while
 
             [
                 dup seq nth-unsafe pivot quot call( x x -- x )
                 +gt+ eq?
-            ] [ 1 - ] while
+            ] [ 1 fixnum-fast ] while
 
             2dup <= [
                 [ seq exchange-unsafe ]
-                [ [ 1 + ] [ 1 - ] bi* ] 2bi
+                [ [ 1 fixnum+fast ] [ 1 fixnum-fast ] bi* ] 2bi
             ] when
         ] while
 
@@ -36,7 +39,7 @@ IN: sorting.quick
 PRIVATE>
 
 : sort! ( seq quot: ( obj1 obj2 -- <=> ) -- )
-    [ 0 over length 1 - ] dip quicksort ;
+    [ 0 over length 1 - ] dip quicksort ; inline
 
 : sort-with! ( seq quot: ( elt -- key ) -- )
     [ compare ] curry sort! ; inline
```

```
IN: scratchpad 100 iota >array gc [ bar ] time
Running time: 3.267524074 seconds
```

**Comment by mrjbq7** [comment]
Fixed in 8e09ee0266455fb49a57d3c2efede785d52c2fa3.
